### ISSUE TITLE
Logging  and JDBC metrics do not work

### ISSUE THREAD

**Issue by santaniello** [stagemonitor/stagemonitor#340]
Hi,  
I set it up my project like the example https://github.com/stagemonitor/spring-petclinic but the loggin and JDBC metrics do not work.

My apllication is Spring-MVC (not Spring Boot).

![captura de tela de 2017-10-28 02-04-06](https://user-images.githubusercontent.com/9118463/32131091-5fbf991e-bb84-11e7-971d-2efbe75ca382.png)

I use log4j and slf4j dependencies.

![captura de tela de 2017-10-28 02-04-14](https://user-images.githubusercontent.com/9118463/32131096-6941a216-bb84-11e7-8e0d-68d20f3b691f.png)

My datasource is a tomcat container and I use jndi to get it in my application. 

```xml
<!-- Data Source     -->
    <Resource name="jdbc/prod" auth="Container" type="javax.sql.DataSource"
              maxActive="50" maxIdle="30" maxWait="10000"
              username="" password=""
              driverClassName="com.mysql.jdbc.Driver"
              url="jdbc:mysql://localhost:3306/teste"/>
```
```java
	@Bean
	@Override
	public DataSource dataSource(){
            JndiDataSourceLookup dataSource = new JndiDataSourceLookup();
	    dataSource.setResourceRef(true);
	    return dataSource.getDataSource("jdbc/prod");	
      }
```

Thanks,
Felipe

**Comment by felixbarny** [comment]
Seem like instrumentation is not working correctly. Are you starting your application with a JDK? Please post the [startup logs](https://github.com/stagemonitor/stagemonitor/wiki/Troubleshooting).

**Comment by santaniello** [comment]
Measurement Session is initialized: [application=spring-archetype] [instance=dev] [host=felipe-Inspiron-3421]
Not sending metrics to InfluxDB (url=null, interval=60s)
Not sending metrics to Elasticsearch (url=[], interval=60s)
No OpenTracing implementation found. Falling back to NoopTracer. This is fine if you just want to use stagemonitor for development, for example with the in-browser-widget. If you want to report your traces to Elasticsearch, add a dependency to stagemonitor-tracing-elasticsearch. If you want to report to Zipkin, add stagemonitor-tracing-zipkin.
Using ConcurrentMapIncidentRepository for storing incidents.
# stagemonitor status
System information: Java 1.8.0_131 (Oracle Corporation) Linux 4.4.0-97-generic
OK   - Agent attachment 
OK   - AlertingPlugin (version 0.84.0)
OK   - CorePlugin (version 0.84.0)
FAIL - Elasticsearch (Elasticsearch is not available)
OK   - JdbcPlugin (version 0.84.0)
OK   - JvmPlugin (version 0.84.0)
OK   - LoggingPlugin (version 0.84.0)
OK   - OsPlugin (version 0.84.0)
OK   - ServletPlugin (version 0.84.0)
OK   - SoapTracingPlugin (version 0.84.0)
OK   - Startup 
OK   - TracingPlugin (version 0.84.0)
stagemonitor configuration, listing non-default values:
stagemonitor.applicationName: spring-archetype (source: stagemonitor.properties)
stagemonitor.instanceName: dev (source: stagemonitor.properties)
stagemonitor.instrument.exclude: br.com.springarchetype.models (source: stagemonitor.properties)
stagemonitor.instrument.include: br.com.springarchetype (source: stagemonitor.properties)
stagemonitor.alerts.frequency: 10 (source: stagemonitor.properties)
stagemonitor.requestmonitor.http.parseUserAgent: true (source: stagemonitor.properties)
stagemonitor.web.widget.enabled: false (source: stagemonitor.properties)
stagemonitor.web.paths.excluded: /resources, /webjars, /dandelion (source: stagemonitor.properties)
stagemonitor.eum.enabled: true (source: stagemonitor.properties)
Registering configuration Endpoint /stagemonitor/configuration. You can dynamically change the configuration by issuing a POST request to /stagemonitor/configuration?key=stagemonitor.config.key&value=configValue&stagemonitor.password=password. If the password is not set, dynamically changing the configuration is not available. The password can be omitted if set to an empty string.
built new end user monitoring JavaScript, size=6394 bytes, etag="113823574"

**Comment by santaniello** [comment]
I'm using this versions of the log4j and slf4j

```xml 

<org.slf4j-version>1.7.7</org.slf4j-version>
<dependency>
	<groupId>org.slf4j</groupId>
	<artifactId>slf4j-api</artifactId>
	<version>${org.slf4j-version}</version>
</dependency>
<dependency>
	<groupId>org.slf4j</groupId>
	<artifactId>jcl-over-slf4j</artifactId>
	<version>${org.slf4j-version}</version>
	<scope>runtime</scope>
</dependency>
<dependency>
	<groupId>org.slf4j</groupId>
	<artifactId>slf4j-log4j12</artifactId>
	<version>${org.slf4j-version}</version>
	<scope>runtime</scope>
</dependency>
<dependency>
	<groupId>log4j</groupId>
	<artifactId>log4j</artifactId>
	<version>1.2.15</version>
	<exclusions>
		<exclusion>
			<groupId>javax.mail</groupId>
			<artifactId>mail</artifactId>
		</exclusion>
		<exclusion>
			<groupId>javax.jms</groupId>
			<artifactId>jms</artifactId>
		</exclusion>
		<exclusion>
			<groupId>com.sun.jdmk</groupId>
			<artifactId>jmxtools</artifactId>
		</exclusion>
		<exclusion>
			<groupId>com.sun.jmx</groupId>
			<artifactId>jmxri</artifactId>
		</exclusion>
	</exclusions>
	<scope>runtime</scope>
</dependency>
		
```

**Comment by felixbarny** [comment]
Which DataSource implementation are you using?

**Comment by jpfourny** [comment]
I am faced with a similar issue, but I am pretty sure it's related to class-loader visibility and timing. In my case, SLF4J and LOG4J are brought into an Websphere Liberty container with a Liberty feature, which is long before Stagemonitor is initialized. Since we are using an IBM JRE, there is little (if any) support for retransformation. Byte buddy never gets to process those classes. :(

**Comment by jpfourny** [comment]
I was thinking of re-packaging stagemonitor (along with byte-buddy) into an uber Java agent that could instrument as early as possible, but this is not an easy feat. One of the technical problems to overcome is dealing with multiple web apps per JVM (in one or more containers). Currently, Stagemonitor class is loaded once for each application (isolated; having their own configurations). By moving initialization earlier, you will end up breaking this model. I am not sure this can be achieved without significant changes to stagemonitor... need to think it through.

**Comment by felixbarny** [comment]
Another challenge is that stagemonitor assumes that it has direct access to classes like `javax.servlet.Filter`, and Spring classes. When attaching as a `javaagent`, stagemonitor's classes would be loaded by the system class loader, which does not have access to those classes.

> I am not sure this can be achieved without significant changes to stagemonitor... need to think it through.

I also think that this would require significant changes.

If you can't force the initialization of stagemonitor earlier (for example with something like https://github.com/stagemonitor/stagemonitor/blob/master/stagemonitor-core/src/main/java/org/stagemonitor/core/instrument/StagemonitorRuntimeAgentAttacherDriver.java) and retransformation is also not available you are out of luck, I'm afraid.

**Comment by jpfourny** [comment]
Yes, good point. I will abandon the agent idea. :D As it turns out, the logger instrumentation is not working for me in a standalone application, either.  That rules out classloader hierarchies being a factor. I can see that instrumentation debugging does acknowledge processing of the Log4j Logger class, but the advice still does not fire. I will keep poking at this.

**Comment by jpfourny** [comment]
I just tried my standalone (non web) app with Oracle JDK and still no luck. I think that rules out IBM-specific issue (ie: retransformation).

**Comment by jpfourny** [comment]
OK, I figured out my specific issue. The instrumentation strategy does not support calling Log4J via SLF4J, because SLF4J takes a backdoor. Here is an example stack:

`
org.slf4j.impl.Log4jLoggerAdapter.info(java.lang.String, java.lang.Object) line: 324	
org.apache.log4j.Logger(org.apache.log4j.Category).log(java.lang.String, org.apache.log4j.Priority, java.lang.Object, java.lang.Throwable) line: 856	
org.apache.log4j.Logger(org.apache.log4j.Category).forcedLog(java.lang.String, org.apache.log4j.Priority, java.lang.Object, java.lang.Throwable) line: 391	
org.apache.log4j.Logger(org.apache.log4j.Category).callAppenders(org.apache.log4j.spi.LoggingEvent) line: 206	
org.apache.log4j.helpers.AppenderAttachableImpl.appendLoopOnAppenders(org.apache.log4j.spi.LoggingEvent) line: 66	
org.apache.log4j.ConsoleAppender(org.apache.log4j.AppenderSkeleton).doAppend(org.apache.log4j.spi.LoggingEvent) line: 251	
org.apache.log4j.ConsoleAppender(org.apache.log4j.WriterAppender).append(org.apache.log4j.spi.LoggingEvent) line: 162	
`

I think a simple fix would be to include `Log4jLoggerAdapter` in MeterLoggingTransformer.getTypeMatcher() (I see JDK14LoggerAdapter is already there). It won't cover all possible logging implementations bridged to SLF4J, but it is an easy addition. Maybe consider instrumenting ALL implementations of org.slf4j.Logger, instead?
